api/api/templates/admin/api/media_report/change_form.html

```diff
@@ -147,4 +147,42 @@ <h2>Other reports</h2>
     img.src = url;
   }
 </script>
+
+<script>
+  // This script tag manages the soft-locking mechanism for the report.
+
+  function toggleLock(url) {
+    // Toggle the soft-lock for the current user-report pair.
+    fetch(url, {
+      method: 'POST',
+      keepalive: true,
+      headers: {
+        'Content-Type': 'application/json',
+        'X-CSRFToken': document.querySelector('[name=csrfmiddlewaretoken]').value
+      },
+    })
+      .then(res => console.log(res))
+      .catch(err => console.error(err));
+  }
+
+  function softLock() {
+    toggleLock('{% url "admin:api_"|add:media_type|add:"report_lock" object_id=object_id %}')
+  }
+
+  function softUnlock() {
+    toggleLock('{% url "admin:api_"|add:media_type|add:"report_unlock" object_id=object_id %}')
+  }
+
+  // On first load, set the soft-lock.
+  document.addEventListener("DOMContentLoaded", softLock)
+  // Subsequent visibility changes (including unload and other ways to
+  // close the page) will toggle the soft-lock.
+  document.addEventListener("visibilitychange", function() {
```

Comment:
This also includes switching tabs, even if the tab stays open. I think that's probably fine for the first pass... but we should make sure it's documented so the moderators understand that nuance here.

If there's a way to account for that in the condition, maybe some property on the event, it'd be good to do so, so that it stays soft locked if the moderator, for example, just opened a tab to review the creator's works on the source, or something like that.

With what we know of how this process will work and who will be doing it (specifically how many people will be doing it), I don't think this is a big deal. If it's easy to account for tab switching, that's great, but if it's a big hassle or introduces other complexities and nuances as a side-effect, then it isn't worth it for this first iteration.

Reply:
In my first implementation, I was using an event just before a page is unloaded, but those approaches are fickle and unreliable. Then I thought that a lock should actually be lifted if the moderator switches to a different tab (considering all the media information is present in the report page itself) but you make a valid point.

I've switched to your polling based approach with short TTL so that even a background tab preserves the lock as long as it remains open.